File: typescript/src/providers/openai/collector.ts
/**
 * Trace collection and serialization for OpenAI.
 */

import type { BaseCollector } from '../base';
import type { Trace, RequestData, ResponseData, ErrorData } from '../../models';
import { getConfig } from '../../config';
import { VERSION } from '../../version';

/**
 * Collects and builds trace objects from OpenAI API calls.
 */
export class OpenAICollector implements BaseCollector {
  getProviderName(): string {
    return 'openai';
  }

  collectTrace(
    traceId: string,
    startTime: number,
    requestArgs: Record<string, unknown>,
    response?: unknown,
    error?: Error
  ): Trace | null {
    try {
      // Calculate duration
      const endTime = Date.now();
      const durationMs = endTime - startTime;

      // Get current timestamp in ISO 8601 UTC format
      const timestamp = new Date().toISOString();

      // Get configuration
      const config = getConfig();
      const environment = config?.environment;

      // Build request data
      const requestData = this.buildRequestData(requestArgs);

      // Build response or error data
      let responseData: ResponseData | undefined;
      let errorData: ErrorData | undefined;

      if (response) {
        responseData = this.buildResponseData(response);
      } else if (error) {
        errorData = this.buildErrorData(error);
      }

      // Build trace in unified format
      const trace: Trace = {
        trace_id: traceId,
        timestamp,
        provider: 'openai',
        method: 'chat.completions.create',
        duration_ms: durationMs,
        request: requestData,
        response: responseData,
        error: errorData,
        sdk_version: `ambertrace-node/${VERSION}`,
        environment,
      };

      if (config?.debug) {
        console.log(`[AmberTrace] Collected OpenAI trace ${traceId} (duration: ${durationMs.toFixed(2)}ms)`);
      }

      return trace;
    } catch (err) {
      // Never raise exceptions - log and return null
      console.error(`[AmberTrace] Failed to collect OpenAI trace ${traceId}:`, err);
      return null;
    }
  }

  private buildRequestData(args: Record<string, unknown>): RequestData {
    // Extract model
    const model = (args.model as string) ?? 'unknown';

    // Extract messages
    const rawMessages = (args.messages as Array<{ role: string; content: string }>) ?? [];
    const messages = rawMessages.map((msg) => ({
      role: msg.role ?? 'unknown',
      content: String(msg.content ?? ''),
    }));

    // Extract parameters (exclude model and messages)
    const parameters: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(args)) {
      if (key !== 'model' && key !== 'messages') {
        parameters[key] = value;
      }
    }

    return {
      model,
      messages,
      parameters,
    };
  }

  private buildResponseData(response: unknown): ResponseData {
    const resp = response as {
      id?: string;
      model?: string;
      choices?: Array<{
        index?: number;
        message?: { role?: string; content?: string };
        finish_reason?: string;
      }>;
      usage?: {
        prompt_tokens?: number;
        completion_tokens?: number;
        total_tokens?: number;
      };
    };

    // Extract response ID
    const id = resp.id ?? 'unknown';

    // Extract model
    const model = resp.model ?? 'unknown';

    // Extract choices
    const rawChoices = resp.choices ?? [];
    const choices = rawChoices.map((choice) => ({
      index: choice.index ?? 0,
      message: {
        role: choice.message?.role ?? 'assistant',
        content: choice.message?.content ?? '',
      },
      finish_reason: choice.finish_reason ?? 'unknown',
    }));

    // Extract usage
    const usage = {
      prompt_tokens: resp.usage?.prompt_tokens ?? 0,
      completion_tokens: resp.usage?.completion_tokens ?? 0,
      total_tokens: resp.usage?.total_tokens ?? 0,
    };

    return {
      id,
      model,
      choices,
      usage,
    };
  }

  private buildErrorData(error: Error): ErrorData {
    const errorType = error.constructor.name;
    const errorMessage = error.message;

    // Try to extract OpenAI error code
    let errorCode: string | undefined;
    const errorAny = error as { code?: string; status?: number; type?: string };

    if (errorAny.code) {
      errorCode = String(errorAny.code);
    } else if (errorAny.status) {
      errorCode = String(errorAny.status);
    } else if (errorAny.type) {
      errorCode = String(errorAny.type);
    }

    return {
      type: errorType,
      message: errorMessage,
      code: errorCode,
    };
  }
}

File: python/src/ambertrace/providers/openai/collector.py
"""Trace collection and serialization for OpenAI."""

import logging
import time
from datetime import datetime, timezone
from typing import Any, Dict, Optional

from ambertrace.config import get_config
from ambertrace.providers.base import BaseCollector

logger = logging.getLogger(__name__)


class OpenAICollector(BaseCollector):
    """Collects and builds trace objects from OpenAI API calls.

    This class is responsible for:
    - Extracting data from OpenAI request/response objects
    - Normalizing to unified dict format
    - Calling build_trace() for dataclass creation and serialization
    - Never raising exceptions (defensive error handling)
    """

    def get_provider_name(self) -> str:
        """Return provider name."""
        return "openai"

    def collect_trace(
        self,
        trace_id: str,
        start_time: float,
        request_kwargs: Dict[str, Any],
        response: Optional[Any] = None,
        error: Optional[Exception] = None,
    ) -> Optional[Dict[str, Any]]:
        """Collect and serialize a trace from an OpenAI API call.

        Args:
            trace_id: Unique trace identifier (UUID)
            start_time: Start timestamp from time.time()
            request_kwargs: Kwargs passed to OpenAI API call
            response: OpenAI response object (if successful)
            error: Exception object (if call failed)

        Returns:
            Serialized trace as dictionary, or None if collection fails
        """
        try:
            # Calculate duration
            end_time = time.time()
            duration_ms = (end_time - start_time) * 1000

            # Get current timestamp in ISO 8601 UTC format
            timestamp = datetime.now(timezone.utc).isoformat()

            # Get configuration
            config = get_config()
            environment = config.environment if config else None

            # Build normalized data dicts
            request_data = self._build_request_data(request_kwargs)
            response_data = None
            error_data = None

            if response is not None:
                response_data = self._build_response_data(response)
            elif error is not None:
                error_data = self._build_error_data(error)

            # Use shared build_trace() for dataclass creation and serialization
            trace_dict = self.build_trace(
                trace_id=trace_id,
                timestamp=timestamp,
                provider="openai",
                method="chat.completions.create",
                duration_ms=duration_ms,
                request_data=request_data,
                response_data=response_data,
                error_data=error_data,
                environment=environment,
            )

            if trace_dict:
                logger.debug(f"Collected OpenAI trace {trace_id} (duration: {duration_ms:.2f}ms)")

            return trace_dict

        except Exception as e:
            # Never raise exceptions - log and return None
            logger.error(f"Failed to collect OpenAI trace {trace_id}: {e}", exc_info=True)
            return None

    def _build_request_data(self, kwargs: Dict[str, Any]) -> Dict[str, Any]:
        """Build request data dict from OpenAI API call kwargs.

        Args:
            kwargs: Keyword arguments passed to OpenAI API

        Returns:
            Normalized request data dict
        """
        # Extract model
        model = kwargs.get("model", "unknown")

        # Extract messages
        raw_messages = kwargs.get("messages", [])
        messages = []
        for msg in raw_messages:
            # Handle both dict and object formats
            if isinstance(msg, dict):
                role = msg.get("role", "unknown")
                content = msg.get("content", "")
            else:
                role = getattr(msg, "role", "unknown")
                content = getattr(msg, "content", "")

            messages.append({"role": role, "content": str(content)})

        # Extract parameters (temperature, max_tokens, etc.)
        # Exclude model and messages from parameters
        excluded_keys = {"model", "messages"}
        parameters = {k: v for k, v in kwargs.items() if k not in excluded_keys}

        return {
            "model": model,
            "messages": messages,
            "parameters": parameters,
        }

    def _build_response_data(self, response: Any) -> Dict[str, Any]:
        """Build response data dict from OpenAI response object.

        Args:
            response: OpenAI ChatCompletion response object

        Returns:
            Normalized response data dict
        """
        # Extract response ID
        response_id = getattr(response, "id", "unknown")

        # Extract model
        model = getattr(response, "model", "unknown")

        # Extract choices
        raw_choices = getattr(response, "choices", [])
        choices = []
        for choice in raw_choices:
            index = getattr(choice, "index", 0)
            finish_reason = getattr(choice, "finish_reason", "unknown")

            # Extract message
            msg = getattr(choice, "message", None)
            if msg:
                role = getattr(msg, "role", "assistant")
                content = getattr(msg, "content", "")
                message = {"role": role, "content": str(content) if content else ""}
            else:
                message = {"role": "assistant", "content": ""}

            choices.append({
                "index": index,
                "message": message,
                "finish_reason": finish_reason,
            })

        # Extract usage
        usage_obj = getattr(response, "usage", None)
        if usage_obj:
            usage = {
                "prompt_tokens": getattr(usage_obj, "prompt_tokens", 0),
                "completion_tokens": getattr(usage_obj, "completion_tokens", 0),
                "total_tokens": getattr(usage_obj, "total_tokens", 0),
            }
        else:
            usage = {"prompt_tokens": 0, "completion_tokens": 0, "total_tokens": 0}

        return {
            "id": response_id,
            "model": model,
            "choices": choices,
            "usage": usage,
        }

    def _build_error_data(self, error: Exception) -> Dict[str, Any]:
        """Build error data dict from exception.

        Args:
            error: Exception that occurred during API call

        Returns:
            Normalized error data dict
        """
        error_type = type(error).__name__
        error_message = str(error)

        # Try to extract OpenAI error code if available
        error_code = None
        if hasattr(error, "code"):
            error_code = str(error.code)
        elif hasattr(error, "status_code"):
            error_code = str(error.status_code)

        return {
            "type": error_type,
            "message": error_message,
            "code": error_code,
        }
